2-virtual-module-source.md

```diff
@@ -0,0 +1,388 @@
+# Virtual Module Sources
+
+## Synopsis
+
+Extend the `Module` constructor such that it accepts virtual module sources:
+objects that implement a protocol that is sufficient for virtualizing the
+evaluation of modules in languages not anticipated by ECMA-262 or a host
+implementations.
+
+## Motivation
+
+Hosts can add support for linking non-EcmaScript languages by creating
+new kinds of ***Module Source Record***.
+For example, a host could add a [[ModuleSource]] internal slot to
+[`WebAssembly.Module`][web-assembly-module] and then link arbitrary WASM into
+an EcmaScript module graph.
+
+However, user code cannot emulate a host in this way.
+To extend or experiment with new languages or new module-like features, user
+code needs a way to emulate or *virtualize* a module source record.
+
+Such a protocol would allow user code to integrate CommonJS, JSON, or WASM
+modules on hosts that do not.
+It would also open a field for experimentation with other kinds of resource
+modules.
+
+## Description
+
+The first argument to the `Module` constructor is a source.
+If that source is an object that does not have a [[Module Source]] internal
+slot, we treat this object as a protocol for loading, binding, linking,
+initializing, and executing the new `Module` instance.
+
+## Interface
+
+```ts
+type VirtualModuleSource = {
+  // Indicates the import and export bindings the module has
+  // between its module environment record, module namespace exotic object,
+  // and its dependencies.
+  bindings?: Array<Binding>,
+
+  // Executes the module if it is imported.
+  // execute may return a promise, indicating that the module uses
+  // the equivalent of top-level-await.
+  execute?: (namespace: ModuleImportsNamespace, {
+    import?: (importSpecifier: string) => Promise<ModuleExportsNamespace>,
+    importMeta?: Object,
+    globalThis?: Object,
```

Comment:
this property should never be missing. and what about the previous idea? that namespace and globalThis merged into one object as the "import lexical scope" + "global scope"

Reply:
@erights felt that reifying the whole module environment record, including its fall-through to the global environment record, coupled too many concerns. I’ll add this to the design questions.

Reply:
It’s probably worth noting that reifying the module environment record in its entirety would put us in a better position to implement the “global lexicals” feature using Evaluators, to implement Compartments.

Reply:
I would be interested in seeing this explored. Frankly, I often forget that we also need to support global lexicals.

Reply:
My understanding is that Moddable’s XS (cc @patrick-soquet @phoddie) reifies the module environment record and I assumed that meant that it reflected the global lexicals and global object of their `Compartment`.

Reply:
It occurs to me that both strategies are viable. We could also surface `globalLexicals` as yet another object in the `execute` arguments bag.

Reply:
What are the global lexicals?

Reply:
Like const/let at the top level of Scripts?

Reply:
I believe let and const at the top level of Scripts are called the “global contour”. There is a separate potential extension for evaluators that would make them more suitable for REPL, where the global contour can be made to persist between Script evaluations. We have not yet figured out whether that’s the same or different than what we’re calling the global lexicals, but it is probably different.

At Agoric, we briefly needed a way to meter a JavaScript program in user code. We eventually found a better way, by subshelling an XS process with an “instructions to process” budget. When we implemented metering in user code, we performed a source transformation that injected meter decrements at junctions. To deny guest code the ability to fiddle with their meter, we censored the name of the meter from source code _and_ needed a place to reveal the variable to object code. The global object was not eligible because any name can be reached by computed key. So we introduced a new “global lexicals” feature to compartments that allowed us to introduce the name *just above* both the top level lexical scope of modules and scripts. There might be other ways to do this or the case might not be sufficiently motivated, but I’m satisfied that there is a way to do this in the future regardless of whether the feature is revealed through a module environment record or a separate argument in the bag.